test-app/tests/integration/components/headless-form-validation-test.gts

```diff
@@ -1622,5 +1754,239 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
+            </form.field>
+            <button type="submit" data-test-submit>Submit</button>
+          </HeadlessForm>
+        </template>);
+
+        await fillIn('[data-test-first-name]', 'Foo');
+
+        assert.false(
+          validateCallback.called,
+          '@validate is not called while typing'
+        );
+
+        await blur('[data-test-first-name]');
+
+        assert.false(
+          validateCallback.called,
+          '@validate is not called until submitting'
+        );
+
+        await click('[data-test-submit]');
+
+        assert.true(
+          validateCallback.calledOnce,
+          '@validate is called when submitting'
+        );
+        assert.true(
+          validateCallback.calledWith('Foo', 'firstName', {
+            ...data,
+            firstName: 'Foo',
+          }),
+          '@validate is called with form data'
+        );
+
+        await input('[data-test-first-name]', 'Tony');
+
+        assert.true(
+          validateCallback.calledTwice,
+          '@validate is called for revalidation on input'
+        );
+        assert.true(
+          validateCallback.getCall(1).calledWith('Tony', 'firstName', {
+            ...data,
+            firstName: 'Tony',
+          }),
+          '@validate is called with form data'
+        );
+
+        await blur('[data-test-first-name]');
+
+        assert.true(
+          validateCallback.calledTwice,
+          '@validate is called for revalidation on focusout'
+        );
+      });
+
+      test('validation errors are exposed as field.errors on input', async function (assert) {
+        const data: TestFormData = { firstName: 'Tony', lastName: 'Foo' };
+
+        await render(<template>
+          <HeadlessForm
+            @data={{data}}
+            @revalidateOn="input"
+            @validate={{validateFormCallbackSync}}
+            as |form|
+          >
+            <form.field @name="firstName" as |field|>
+              <field.label>First Name</field.label>
+              <field.input data-test-first-name />
+              <field.errors data-test-first-name-errors />
+            </form.field>
+            <form.field @name="lastName" as |field|>
+              <field.label>Last Name</field.label>
+              <field.input data-test-last-name />
+              <field.errors data-test-last-name-errors />
+            </form.field>
+            <button type="submit" data-test-submit>Submit</button>
+          </HeadlessForm>
+        </template>);
+
+        assert
+          .dom('[data-test-first-name-errors]')
+          .doesNotExist(
+            'validation errors are not rendered before initial validation happens before form is filled in'
+          );
+        assert
+          .dom('[data-test-last-name-errors]')
+          .doesNotExist(
+            'validation errors are not rendered before initial validation happens before form is filled in'
+          );
+
+        await fillIn('[data-test-first-name]', 'Foo');
```

Comment:
Do we want `input` here or `fillIn`?

Reply:
I think @simonihmig is doing something similar here as the above tests - ensuring `fillIn` _doesn't_ cause errors to render, but below on line 1976 where he's using `input` it does

Reply:
I guess it doesn't matter in this case too because the revalidate doesn't start until after `submit`

Reply:
Yeah, it doesn't matter here. And `fillIn` is basically doing more as explained elsewhere (triggering `input` _and_ `change` events), so when even `fillIn` does not make errors show up, then `input` would neither.